src/modules-load/modules-load-generator.c

```diff
@@ -0,0 +1,131 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "conf-files.h"
+#include "constants.h"
+#include "errno-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "generator.h"
+#include "log.h"
+#include "proc-cmdline.h"
+#include "special.h"
+#include "string-util.h"
+#include "strv.h"
+#include "unit-name.h"
+
+static const char *arg_dest = NULL;
+
+static char **arg_proc_cmdline_modules = NULL;
+static const char conf_file_dirs[] = CONF_PATHS_NULSTR("modules-load.d");
+
+STATIC_DESTRUCTOR_REGISTER(arg_proc_cmdline_modules, strv_freep);
+
+static int parse_proc_cmdline_item(const char *key, const char *value, void *data) {
+        int r;
+
+        if (proc_cmdline_key_streq(key, "modules_load")) {
+
+                if (proc_cmdline_value_missing(key, value))
+                        return 0;
+
+                r = strv_split_and_extend(&arg_proc_cmdline_modules, value, ",", /* filter_duplicates = */ true);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse modules_load= kernel command line option: %m");
+        }
+
+        return 0;
+}
+
+static int add_module_load(const char *name) {
+        _cleanup_free_ char *e = NULL;
+        char *p;
+
+        p = strdupa_safe(name);
+        if (!p)
+                return log_oom();
+
+        e = unit_name_escape(string_replace_char(p, '-', '_'));
+        if (!e)
+                return log_oom();
+
+        return generator_add_symlink_full(arg_dest, SPECIAL_MODULES_LOAD_TARGET, "wants",
+                                          SYSTEM_DATA_UNIT_DIR "/modprobe@.service", e);
+}
+
+static int add_modules_from_file(const char *path) {
+        _cleanup_fclose_ FILE *f = NULL;
+        _cleanup_free_ char *pp = NULL;
+        int r;
```

Comment:
Please declare `int ret = 0, r;`, then use `r` for storing result of each function call, and use `ret` for gathering results.